mgmtd/mgmt_db.c

```diff
@@ -0,0 +1,1040 @@
+/*
+ * MGMTD Databases
+ * Copyright (C) 2021  Vmware, Inc.
+ *		       Pushpasis Sarkar <spushpasis@vmware.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+#include "md5.h"
+#include "mgmtd/mgmt.h"
+#include "mgmtd/mgmt_memory.h"
+#include "mgmtd/mgmt_db.h"
+#include "mgmtd/mgmt_trxn.h"
+#include "libyang/libyang.h"
+
+#ifdef REDIRECT_DEBUG_TO_STDERR
+#define MGMTD_DB_DBG(fmt, ...)                                                 \
+	fprintf(stderr, "%s: " fmt "\n", __func__, ##__VA_ARGS__)
+#define MGMTD_DB_ERR(fmt, ...)                                                 \
+	fprintf(stderr, "%s: ERROR, " fmt "\n", __func__, ##__VA_ARGS__)
+#else /* REDIRECT_DEBUG_TO_STDERR */
+#define MGMTD_DB_DBG(fmt, ...)                                                 \
+	do {                                                                   \
+		if (mgmt_debug_db)                                             \
+			zlog_err("%s: " fmt, __func__, ##__VA_ARGS__);         \
+	} while (0)
+#define MGMTD_DB_ERR(fmt, ...)                                                 \
+	zlog_err("%s: ERROR: " fmt, __func__, ##__VA_ARGS__)
+#endif /* REDIRECT_DEBUG_TO_STDERR */
+
+struct mgmt_db_ctxt {
+	Mgmtd__DatabaseId db_id;
+	pthread_rwlock_t rw_lock;
+
+	bool config_db;
+
+	union {
+		struct nb_config *cfg_root;
+		struct lyd_node *dnode_root;
+	} root;
+};
+
+struct mgmt_cmt_info_t {
+	struct mgmt_cmt_info_dlist_item cmt_dlist;
```

Comment:
What exactly is this, why is it called a database, and why is it not using something like SQLite, GDBM or LMDB?

Reply:
This is a linked list of commit buffers (upto 10 maximum, with oldest entry removed after 10). Entries in each entry of this list points to a file on disk that has a YANG/JSON based snapshot of the running DB after a specific commit. If not clear already we are supporting commit rollback functionality with capability to rollback upto 10 last commits. Hope it answers your question.

Reply:
Sure we will not call the commit history buffer any kind of database.

The commit buffers are written out to the a file at the end of each commit. Also the current state of running DB is written out to the disk at the end of commit. There's a very small window between applying the configs and writing it out to the disk. If power goes out in between, the router will restart back from the older config already present on the disk. The recent commit will be lost. Maybe in al later PR we can come up with a solution for that too. But at the moment don't want to complicate the implementation further.